test/coretasks/test_coretasks_sasl.py

```diff
@@ -253,6 +253,26 @@ def test_sasl_plain_no_password(
     ), 'No password is a configuration error and the bot must quit.'
 
 
+# https://github.com/sopel-irc/sopel/issues/2560
+def test_sasl_external_no_password(
+    botfactory: BotFactory,
+    configfactory: ConfigFactory,
+) -> None:
+    tmpconfig = configfactory("conf.ini", TMP_CONFIG_SASL_NO_PASSWORD)
+    mockbot = botfactory.preloaded(tmpconfig, preloads=["coretasks"])
+    mockbot.settings.core.auth_target = "EXTERNAL"
+    mockbot.backend.connected = True
+
+    # connect and capability negotiation
+    mockbot.on_connect()
+    mockbot.on_message(":irc.example.com CAP * LS :sasl=PLAIN,EXTERNAL")
+    n = len(mockbot.backend.message_sent)
+    mockbot.on_message(":irc.example.com CAP * ACK :sasl")
+    assert mockbot.backend.message_sent[n:] == rawlist(
+        "AUTHENTICATE EXTERNAL"
+    ), "SASL EXTERNAL did not continue without a password configured."
```

Comment:
Since this is not a fail state, I'd like to see the test to do the same thing as the SASL plain test, i.e. see the server sends a "all done" so the bot can reply with a CAP END, like so (I'm pretty sure this should pass as-is):

```python
    # Server accept authentication
    mockbot.on_message(
        ':irc.example.com 900 TestBot TestBot!sopel@example.com sopel '
        ':You are now logged in as TestBot')
    mockbot.on_message(
        ':irc.example.com 903 TestBot :SASL authentication successful')
    assert mockbot.backend.message_sent[n:] == rawlist(
        'CAP END',
    )
```

This we know that if the external mechanism works as intended between Sopel and the server without explicit interaction at the IRC protocol level, then we have proof that this doesn't stop Sopel from sending the requested CAP END.

Reply:
I don't believe SASL EXTERNAL is compliant if it proceeds without an exchange of `AUTHENTICATE +`. See raw logs from my testing of this patch as it existed at the tip of #2362 last night:

```
2023-11-12 01:10:31,825 <<      ':irc.rizon.io CAP SopelTest ACK :sasl\r\n'
2023-11-12 01:10:31,825 >>      'AUTHENTICATE EXTERNAL\r\n'
2023-11-12 01:10:31,885 <<      'AUTHENTICATE +\r\n'
2023-11-12 01:10:31,886 >>      'AUTHENTICATE +\r\n'
2023-11-12 01:10:31,948 <<      ':irc.rizon.io 900 SopelTest SopelTest!sopel@my-host.name SopelTest :You are now logged in as SopelTest.\r\n'
2023-11-12 01:10:31,949 <<      ':irc.rizon.io 903 SopelTest :SASL authentication successful\r\n'
2023-11-12 01:10:31,950 >>      'CAP END\r\n'
```

Then again, I could be completely off-base. Actual examples of the EXTERNAL mechanism are really hard to find; the IRCv3 spec pages say that EXTERNAL is one of the defined mechanisms, but only gives example client-server exchanges for PLAIN or SCRAM-SHA-1. Without diving into IRCd source code, the easiest way to investigate this is to observe how other implementations behave.

I didn't bother registering a certificate fingerprint at Libera, but it too performs a round of `AUTHENTICATE +` as part of the EXTERNAL mechanism:

```
2023-11-12 17:44:01,017 <<      ':molybdenum.libera.chat CAP SopelTest ACK :sasl\r\n'
2023-11-12 17:44:01,018 >>      'AUTHENTICATE EXTERNAL\r\n'
2023-11-12 17:44:02,148 <<      'AUTHENTICATE +\r\n'
2023-11-12 17:44:02,149 >>      'AUTHENTICATE +\r\n'
2023-11-12 17:44:03,232 <<      ':molybdenum.libera.chat 904 SopelTest :SASL authentication failed\r\n'
2023-11-12 17:44:03,234 >>      'CAP END\r\n'
2023-11-12 17:44:03,234 >>      'QUIT :SASL Auth Failed\r\n'
```

(Testing 904 w/EXTERNAL is probably a good test case, too, btw.)